boot/bootutil/src/loader.c

```diff
@@ -715,10 +743,10 @@ boot_validate_slot(int slot, struct boot_status *bs)
     }
 
     hdr = boot_img_hdr(&boot_data, slot);
-    if (boot_magic_is_erased(flash_area_erased_val(fap), hdr->ih_magic) ||
-            hdr->ih_flags & IMAGE_F_NON_BOOTABLE) {
+    if (boot_check_header_erased(slot) == 0 || (hdr->ih_flags & IMAGE_F_NON_BOOTABLE)) {
         /* No bootable image in slot; continue booting from slot 0. */
-        return -1;
+        rc = -1;
+        goto out;
```

Comment:
IMO it is better to have several calls to `flash_area_close` than to introduce `goto`.
It uses the same amount of code lines, and has a simpler code flow to read, since the return is immediately afterwards, and one can jump to the next function.

Reply:
This might be true right now. But if there were other clean up steps, it would not be true anymore, and instead of changing everything one would just insert the new clean up step at the exit label. One reason why this is such a common pattern in C. As a matter of style I actually like the way it is done now.